### PR TITLE
Change all jsonmerge to strategicpatch

### PR DESCRIPTION
**Pull request by ping035627**
Implement all jsonmerge changed to strategicpatch in edit.go.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32274)

### CODE REVIEW

File: pkg/util/strategicpatch/patch.go [hunk 1]
```diff
@@ -110,6 +110,28 @@ func RequireKeyUnchanged(key string) PreconditionFunc {
 	}
 }
 
+// RequireMetadataKeyUnchanged creates a precondition function that fails
+// if the metadata.key is present in the patch (indicating its value
+// has changed).
+func RequireMetadataKeyUnchanged(key string) PreconditionFunc {
```
Comment (deads2k) [inline pkg/util/strategicpatch/patch.go]: This precondition seems wrong.  Annotations and Labels are in metadata and I should be allowed to change them.

Comment (deads2k) [inline pkg/util/strategicpatch/patch.go]: > This precondition seems wrong. Annotations and Labels are in metadata and I should be allowed to change them.

nm, re-reading doc/name it make sense.

File: pkg/kubectl/cmd/edit.go [hunk 2]
```diff
@@ -392,25 +391,20 @@ func RunEdit(f *cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args
 					return nil
 				}
 
-				patch, err := strategicpatch.CreateStrategicMergePatch(originalJS, editedJS, currOriginalObj)
-				// TODO: change all jsonmerge to strategicpatch
-				// for checking preconditions
-				preconditions := []jsonmerge.PreconditionFunc{}
+				preconditions := []strategicpatch.PreconditionFunc{}
+				preconditions = append(preconditions, strategicpatch.RequireKeyUnchanged("apiVersion"))
```
Comment (deads2k) [inline pkg/kubectl/cmd/edit.go]: construct inline please.

**Comment by deads2k** [comment]
nit.  Looks like a good patch, thanks.

**Comment by ping035627** [comment]
@deads2k  I have modified, please review again, thanks.

**Comment by ping035627** [comment]
@k8s-bot test this issue: #32441

**Comment by ping035627** [comment]
@deads2k  can you help to trigger a new test? It should be a flake like #32441, thanks.

**Comment by deads2k** [comment]
@k8s-bot test this: issue #32441

**Pull request merged**